Merge Win UI.extensions

tajbender opened the pull request.

tajbender commented:
🚶 Coming soon.

⚠️ Work in Progress

tajbender commented on WinUI.Extensions/Vanara.WinUI.Extensions.csproj:
We'll have to delete 

- TargetFrameworks net5.0-windows

tajbender replied:
No, we don't have to. Minimum should be net5

tajbender replied:
https://github.com/dahall/Vanara/pull/507#issuecomment-2713830381

dahall replied:
There are a number of assemblies that don't support all the same .NET versions as the rest of the project. .NET 5-9 and maybe Core 3.1 would be fine.

tajbender replied:
Thanks for clarifying 👍🏻

tajbender commented:
@dahall One question, when pushing a integration test version of WinUI package, will it be included when the merge is still in progress?

I guess not, is it?

Regards, tajbender

dahall commented:
> @dahall One question, when pushing a integration test version of WinUI package, will it be included in the nightly AppVoyer build when the merge is still in progress?

I don't think so. Also, I have migrated away from AppVeyor and now use GitHub Actions to build after each commit and then post to MyGet.

tajbender commented:
> I don't think so. Also, I have migrated away from AppVeyor and now use GitHub Actions to build after each commit and then post to MyGet.

That's fine for me. I'm asking cause I crashed my dev setup for any reason, nowadays i am not even able to compile my last revisions I checked in 🐞 on the current Win 11 Insider.

However, I restarted from scratch with the latest iterations of Win App SDK with focus on `ExplorerBrowser` control alone. Another question is targeting the `Windows Community Toolkit`.

I use some of their components within the codebase, but I guess _*you don't want*_ such dependencies for  Vanara. That's - however - not a show-stopper. I just would copy & paste i.e. import their classes into the package itself. It's not that much classes I use, but I use some especially for performance reasons.

e.g.: https://github.com/CommunityToolkit/Windows/tree/main/components/Collections/src/AdvancedCollectionView

@dahall Is that okay to you?

dahall commented:
> @dahall Is that okay to you?

This all sounds good. I stay holding until I hear you are ready to integrate.

tajbender commented:
Thanks for your patience. I won't touch any other packages, nor do I need to for that stuff.

I'm living on the WinUI package island solely 😃 It's completely decoupled from the existing Vanara code base, I just use your Shell stuff.

The biggest Issue will be that namespaces and classes don't collide - beside code quality, for sure.

dahall commented:
Curious to have you try the 4.1.0 packages. I added .net 9.0 support and removed 4.5. Most notable is that all handles are now generated. I'm trying my hand at generation before using in many more places and maybe even pulling in generated code from https://github.com/microsoft/win32metadata to recreate Vanara live like CsWin32, but with the Vanara primitives and design goals.

tajbender commented:
Thank you so much. Without doubt, the next days I finally _will_ submit code.

As mentioned yesterday, [NETSDK1138: The target framework 'net7.0-windows10.0.19041.0' is out of support..](https://github.com/dahall/Vanara/issues/506#issuecomment-2703024416) made me crazy for weeks now, cause every then and when this error came and left, and I don't get the point.

Then I tried JetBrains Rider, and _*that beast*_ doesn't even mention there _could_ be a problem and compiles fine.

_I'm definitely too old for this_ 💩 - But I'm fine with the lessons learned, another year, another try 🧨. Now, I switched [my dev-branch to reconfigure from scratch](https://github.com/electrifier/electrifier-v1.25/tree/reconfigure) and ignore the .net7 issue. The XAML is already in, code stubs are coming. Everything works and compiles without this solely 'issue', but I don't even think it is one.

Long writing, short story: Now comes the fun part again, _writing code_ 🎃

tajbender commented:
@dahall Surprise!

I'm very happy to let you know that we'll not only have a full working clone of `ExplorerBrowser`, it will also add double pane mode by default, without any hassle 🎈 

![image](https://github.com/user-attachments/assets/0e5da11d-8f2e-41c9-aa86-f48ab9b15d11)

tajbender commented:
if'd like to take a flight: https://github.com/electrifier/electrifier-v1.25.git ⚠️

tajbender commented:
> Curious to have you try the 4.1.0 packages. I added .net 9.0 support and removed 4.5. Most notable is that all handles are now generated. I'm trying my hand at generation before using in many more places and maybe even pulling in generated code from https://github.com/microsoft/win32metadata to recreate Vanara live like CsWin32, but with the Vanara primitives and design goals.

fyi, I'm taking the direct route 🫖 

![image](https://github.com/user-attachments/assets/9787eda7-c2a5-464e-856c-3e5ed05ae8a8)

tajbender commented:
See WinUI 3 official Requirement: https://github.com/dahall/Vanara/pull/494#issuecomment-2595456933

tajbender commented:
That is the warning, I currently get:

```
Updating the layout...
Copying files: Total <1 mb to layout...
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 09:55 and took 02.814 seconds ==========
========== Deploy: 1 succeeded, 0 failed, 0 skipped ==========
========== Deploy completed at 09:55 and took 02.814 seconds ==========
DEP0840: Project 'electrifier' requires packages 'MicrosoftCorporationII.WinAppRuntime.Main.1.4_8wekyb3d8bbwe' and 'MicrosoftCorporationII.WinAppRuntime.Singleton_8wekyb3d8bbwe' with version >= '4000.1010.1349.0'.
These packages should be installed on first run, if not please visit http://go.microsoft.com/fwlink/?linkid=2222757 for installation instructions.
Deployment complete (0:00:00.061). Full package name: "01234567-affe-0101"
```

tajbender commented:
Certainly! Developing a Windows app using the Windows App SDK with C# requires a few key components and steps. Here's a concise overview to get you started:

System Requirements
Operating System: Windows 10 version 1809 (build 17763) or later.
Development Environment: Visual Studio 2019 version 16.9 or later, or Visual Studio 2022.
.NET SDK: .NET 5.0 or later.
Software Requirements
Windows App SDK: Install the latest version of the Windows App SDK.
Visual Studio Workloads:
Universal Windows Platform (UWP) development: Required for UWP apps.
Desktop development with C#: Required for WinUI 3 desktop apps.
Steps to Set Up Your Development Environment

Install Visual Studio:

Download and install Visual Studio from the [official website](https://visualstudio.microsoft.com/).
During installation, select the necessary workloads mentioned above.

Install .NET SDK:

Download and install the .NET SDK from the [.NET website](https://dotnet.microsoft.com/download).

Install Windows App SDK:

Use the Visual Studio Installer to add the Windows App SDK extension.
Alternatively, you can install it via NuGet in your project.
Creating a New Project

Open Visual Studio.

Create a new project:

Select Create a new project.
Choose a template such as Blank App, Packaged (WinUI 3 in Desktop) or Blank App, Unpackaged (WinUI 3 in Desktop).
Click Next and configure your project settings.

Configure Project Dependencies:

Ensure your project references the Windows App SDK.
Add necessary NuGet packages for WinUI 3 and other dependencies.
Example Code Snippet

Here's a simple example of a WinUI 3 app in C#:

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace MyWinUIApp
{
    public sealed partial class MainWindow : Window
    {
        public MainWindow()
        {
            this.InitializeComponent();
            var textBlock = new TextBlock
            {
                Text = "Hello, Windows App SDK!",
                VerticalAlignment = VerticalAlignment.Center,
                HorizontalAlignment = HorizontalAlignment.Center
            };
            this.Content = textBlock;
        }
    }
}


This example creates a basic window with a centered text block displaying a greeting message.

Additional Resources
[Windows App SDK Documentation](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/)
[WinUI 3 Documentation](https://docs.microsoft.com/en-us/windows/apps/winui/winui3/)

Feel free to reach out if you need more detailed guidance or have specific questions!

tajbender commented:
@dahall I upgraded to WinAppSDK 1.5, `1.5.25 - latest stable` a far as i remember. Just an update.

tajbender commented:
Here is what we would get, at least and so far:

![image](https://github.com/user-attachments/assets/13620a26-5e1c-42d8-ae23-04d7ef47f31a)

The complete `Controls` namespace as seen here would be merged into the Vanara package, plus, at least a Breadcrumb Bar.

The Highlighted Ressource file will contain all i18n aware texts, that we'll use especially for Commands and Error / Info messages. I myself could provide German and English, obviously.

tajbender commented:
Just a Note: See https://github.com/CommunityToolkit/Windows/issues/533#issuecomment-2475033872

I had to target Windows version 22621 at least, to get the Windows Community Toolkit running:

> When consuming the Toolkit, you must target Windows version 22621 instead of the prior 18362, i.e. if you were targeting net6.0-windows10.0.18362, you would target net6.0-windows10.0.22621 instead now.
> This does not affect your ability to run down-level with the minimum target version, that remains at 17763.

tajbender commented:
https://github.com/dahall/Vanara/pull/593
